kube-runtime/src/watcher.rs

```diff
@@ -233,11 +234,22 @@ where
                 (Some(Err(err).map_err(Error::InitialListFailed)), State::Empty)
             }
         },
-        State::InitListed { resource_version } => match api.watch(list_params, &resource_version).await {
-            Ok(stream) => (None, State::Watching {
-                resource_version,
-                stream,
-            }),
+        State::InitListed { resource_version } => match {
+            // Watch requests do not support setting resource_version and resource_version_match.
+            // This is a hack until we split List and Watch params to make watcher able to benefit from setting resource versions for list requests.
+            let mut watch_params = list_params.clone();
+            watch_params.resource_version = None;
+            watch_params.resource_version_match = None;
```

Comment:
This looks like it will fix the issue at the cost of more cloning, but the `ListParams` struct is barely holding itself together as it stands.

We could split listparams into {list,watch}params, but then the `watcher`s either:
1. needs to take both params (forcing the user to duplicate selectors and stuff awkwardly)
2. needs a specific parameter struct (watcher::Config?) that contains the common subset of options

I'm thinking the second option makes more sense here. The watch/list params split will still untangle the original problem and simplifies validation of the structs, and a new config for watcher allows concise stream usage to better expose the better behaviour.

quick ideas for a new `watcher::Config`:
- contains most listparam options but excludes resource version and continue tokens (for now, we don't do pagination in watcher atm)
- could implement `TryInto<ListParams>` and `TryInto<WatchParams>`

have you got opinions here @nightkr ?

Reply:
I think that makes sense to do. I'd probably do private methods rather than `TryInto`, so that we don't make it a part of the public API (yet, anyway).

Reply:
It seems reasonable to me. I will try to code it today.

Reply:
@nightkr @clux I refactored the code. The pars with ListParams / WatchParams is much cleaner now in terms of validating.